### ISSUE TITLE
PublishSubject<E> is marked as final

### ISSUE THREAD

**Issue by shahan312** [ReactiveX/RxSwift#270]
PublishSubject<E> is marked as final in beta3, preventing the ability to subclass it. Could we get the final keyword removed?

thanks,
shahan

**Comment by kzaher** [comment]
Hi @shahan312 ,

Could you maybe explain the context why do you need that feature :)

As far as I can tell, it's `final` in other implementations .... that support `final`.
RxJs
```
public final class PublishSubject<T> extends Subject<T, T> {

    /**
     * Creates and returns a new {@code PublishSubject}.
     *
     * @param <T> the value type
     * @return the new {@code PublishSubject}
     */
```

Rx.NET
```
/// <summary>
    /// Represents an object that is both an observable sequence as well as an observer.
    /// Each notification is broadcasted to all subscribed observers.
    /// </summary>
    /// <typeparam name="T">The type of the elements processed by the subject.</typeparam>
    public sealed class Subject<T> : SubjectBase<T>, IDisposable
    {
```

You can easily create a composition of `PublishSubject` in your code if needed :) It has practically only one or two methods :)

**Comment by shahan312** [comment]
Hi @kzaher,

We're subclassing it to implement the ReactJs flux pattern. Here's an example:

```swift
// Store is a subclass of PublishSubject with additional methods specific to its function
// We also have other subclasses for other components from the flux pattern
class Store<Action>: PublishSubject<Action> {
    ...
}

enum StateStoreAction {
    case Login
    ....
    case Logout
}

class StateStore: Store<StateStoreAction> {
    static let sharedInstance = StateStore()
    ...
}

// elsewhere, ie. in a viewDidLoad or something
StateStore.sharedInstance.subscribeNext { action in
    ...
}.addTo(disposable);
```

Please let me know your recommendations.

**Comment by dtartaglia** [comment]
How about this?

    class Store<Action> {
        let subject: PublishSubject<Action>
        ...
    }

    enum StateStoreAction {
        case Login
        ...
        case Logout
    }

    class StateStore: Store<StateStoreAction> {
        static let sharedInstance = StateStore()
        ...
    }

    // elsewhere, i.e. in a viewDidLoad or something
    StateStore.sharedInstance.subject.subscribeNext { action in 
        ...
    }.addTo(disposable);

If you don't like the extra indirection, you can put a `subscribeNext` in Store that simply routes the call to its `subject` object.

In other words, I don't see anything in your example that would require you to subclass PublishSubject...

**Comment by kzaher** [comment]
@shahan312 yep, think @dtartaglia answered.

I think `Variable` https://github.com/ReactiveX/RxSwift/blob/master/RxSwift/Subjects/Variable.swift#L16 is an example of subject composition inside Rx. You can do the same for `class Store<Action>`.

**Comment by shahan312** [comment]
Sounds good. I'll switch over to using subject composition :) Thank you so much @dtartaglia and @kzaher.
